### PR TITLE
Change RetrieveCertificate so that it resets the enrollment (if it exists) while retrieving the certificate

### PR DESCRIPTION
**Pull request by maelvls**
**Update 29 Nov 2022:** I put this PR back to "draft" because I am less and less confident with the changes I am making here. I propose to first agree on a set of `RetrieveCertificate` tests in #270 before I make any change to it `RetrieveCertificate`. When we have merged #270, we can proceed with the current PR.
**Update 5 Dec 2022:** I am now confident with the PR, it can now be reviewed.

---

**The problem:** When enrolling a new certificate, for example by running `vcert enroll` or when using cert-manager, people get "stuck" with an error of the like:

    500 Certificate \VED\Policy\Test\foo.com has encountered an error while processing,
    Status: This certificate cannot be processed while it is in an error state. Fix any
    errors, and then click Retry., Stage: 700.

This message occurs when a past enrollment has failed or an enrollment was still in progress for that certificate. The current workaround is to call to `POST /reset` with Restart=False, and then re-run the command `vcert enroll` (or renew the certificate in cert-manager).

This PR implements the design presented in [Solution 3: request, then reset if retrieve returns “Click Retry” or “WebSDK CertRequest”](https://hackmd.io/@maelvls/design-reset-cert-when-requesting#Solution-3-request-then-reset-if-retrieve-returns-%E2%80%9CClick-Retry%E2%80%9D-or-%E2%80%9CWebSDK-CertRequest%E2%80%9D). The OAuth token doesn't need to change (`certificate:manage`).

The [Jenkins build 2378](https://jenkins.eng.venafi.com/blue/organizations/jenkins/VCert%2Fvcert/detail/vcert/2378/pipeline) is passing for 63f7dff.

**Self-review:**

- I went with a "mock" HTTP server due to the difficulty to get the 202 and 500 HTTP status codes "on demand" from TPP. I didn't add a "live test" because I don't know how to consistently force TPP to return a 500 without RDP'ing into the VM and putting a PowerShell script or turning off the Windows CA (it can't be a policy check such as the domain, as this would be a "Stage 0" error which is the only stage number which gets reset upon requesting a new certificate).
  - **Fake server:** use a fake server to test it, but there is currently no fake server (note that @wallrj started to write a fake server in #262). The fake server, as opposed to the mock server, contains some logic.
  - **Mock responses:** use mock HTTP responses with a mock HTTP server.

**Manual test performed:**

Apart from the "mock" tests that I added, I also manually tested this change with a live TPP instance:

<details>
<summary>Reproducing the manual test</summary>

1. I spun up a TPP instance on CloudShare (using my Jetstack account), using the template "Master NGINX/K8s Blueprint" (snapshot "[6] training_2020-08-28") that comes with TPP 20.1.
2. I went to https://uvo1dq3vmkecfydcwxm.env.cloudshare.com/aperture/application-integrations/ and create a new API integration, since none of the existing contains `configuration:manage` (which I need to set the correct policy):
  
4. I created a token and stored it in the env var `TOKEN` with the following:
  
    ```sh
    TPP_URL=https://uvo1dq3vmkecfydcwxm.env.cloudshare.com
    TPP_USER=tppadmin
    TPP_PASSWORD=
    TOKEN=$(vcert getcred -u $TPP_URL --username $TPP_USER --password $TPP_PASSWORD --client-id=test --scope='certificate:manage;configuration:manage' --format json | jq -r .access_token)
    ```

5. Since we need to trigger an enrollment failure at a stage higher than 0, we will purposefully break the CA. To do that, I RDP'ed into the TPP VM, I opened the program `pkiview`, right-clicked on the only entry, "Manage CA...", and then clicked the "Stop" button:

    

6. I then enrolled:
  
    ```sh
    vcert enroll -u "$TPP_URL" -t "$TOKEN" --cn example.com -z 'TLS/SSL' --san-dns=example.com
    ```

    As expected, it shows:

    > vCert: 2022/11/14 14:58:38 unable to retrieve: Unexpected status code on TPP Certificate Retrieval. Status: 500 Certificate \VED\Policy\TLS/SSL\example.com has encountered an error while processing, Status: Post CSR failed with error: Cannot connect to the certificate authority (CA). Verify that your CA template settings are correct and that the remote server is available. For more information, search the Help system for Configuring the Microsoft Certificate Services Template Object., Stage: 500.

7. If we try to submit the certificate again, we still get the error at stage 500 (weirdly, the original disappeared and a generic message appears instead):
  
    > vCert: 2022/11/14 14:59:43 unable to retrieve: Unexpected status code on TPP Certificate Retrieval. Status: 500 Certificate \VED\Policy\TLS/SSL\example.com has encountered an error while processing, Status: This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry., Stage: 500.

8. I then fixed the CA:
  
    
   

9. I then tried to enroll again, and the "old" error with the sage 500 still shows:
  
    > vCert: 2022/11/14 15:06:30 unable to retrieve: Unexpected status code on TPP Certificate Retrieval. Status: 500 Certificate \VED\Policy\TLS/SSL\example.com has encountered an error while processing, Status: This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry., Stage: 500.

10. Now, from this project's folder, let us run `vcert` with the changes made in this PR:
    
    ```sh
    git fetch origin refs/pull/269/head:reset-before-request
    git checkout reset-before-request
    go run ./cmd/vcert enroll -u "$TPP_URL" -t "$TOKEN" --cn example.com -z 'TLS/SSL' --san-dns=example.com
    ```
  
    This time, the certificate gets properly requested.

</details>

**Changes in error messages:** this PR purposefully doesn't change any of the error messages.

**Possible breakages to other projects:**

- ✅  [terraform-provider-venafi](https://github.com/Venafi/terraform-provider-venafi) uses RetrieveCertificate in three locations. [One](https://github.com/Venafi/terraform-provider-venafi/blob/5374fa5ae86bb09609d228ef89f90a89261d3d3a/venafi/resource_venafi_certificate.go#L818) of them matches on the following specific error message:
    ```go
    "unable to retrieve: Unexpected status code on TPP Certificate Retrieval. Status: 400 Failed to lookup private key, error: Failed to lookup private key vault id"
     ```
     I added a test proving that this same error message will be shown in the same ciscumstances.
- ✅  [vault-pki-backend-venafi](https://github.com/Venafi/vault-pki-backend-venafi) won't be affected since it doesn't parses messages. It uses `RetrieveCertificate` [once](https://github.com/Venafi/vault-pki-backend-venafi/blob/88b14567c280d3d1b57addc2e51d4b5764e7b64e/plugin/pki/vcert_test.go#L61).
- ✅  [cert-manager](https://github.com/cert-manager/cert-manager) has a [single call](https://github.com/cert-manager/cert-manager/blob/d85e424cd0741bdb0fd4a5837d376c690e80fb9f/pkg/issuer/venafi/client/request.go#L65) to `RetrieveCertificate` and doesn't look at the messages.

### CODE REVIEW

File: pkg/venafi/tpp/connector_test.go [hunk 1]
```diff
@@ -51,29 +51,7 @@ var ctx *test.Context
 
 func init() {
 	ctx = test.GetEnvContext()
-	http.DefaultTransport.(*http.Transport).TLSClientConfig = &tls.Config{
```
Comment (inteon) [inline pkg/venafi/tpp/connector_test.go:54]: Why did you remove this code?

File: pkg/venafi/tpp/connector_test.go [hunk 2]
```diff
@@ -528,6 +506,73 @@ func TestRequestCertificateWithValidHours(t *testing.T) {
...
+		switch {
+		case r.URL.Path == "/vedsdk/Identity/Self":
+			_, _ = w.Write([]byte("{}"))
+		case r.URL.Path == "/vedsdk/certificates/retrieve":
+			req := struct{ CertificateDN string }{}
+			_ = json.NewDecoder(r.Body).Decode(&req)
+			if req.CertificateDN != `\VED\Policy\Foo\test.com` {
+				t.Fatalf("/vedsdk/certificates/retrieve: expected CertificateDN to be %s but got %s", `\VED\Policy\Foo\test.com`, req.CertificateDN)
+			}
+			w.WriteHeader(500)
```
Comment (inteon) [inline pkg/venafi/tpp/connector_test.go:540]: Would it be useful/ possible to also test a 202 case or a 404 case?

File: pkg/venafi/tpp/connector_test.go (hunk 1 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector_test.go:54]: Oops. That is my mistake, I will re-add this bit.

File: pkg/venafi/tpp/connector_test.go (hunk 2 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector_test.go]: Yes. I have added three other test cases: one with 200, one with 202, and one with an "unexpected" 500.

Comment (inteon) [reply pkg/venafi/tpp/connector_test.go]: For context: the "unexpected" 500 is tested in the "should not reset when /certificates/retrieve returns an unrelated 500" function (I think)

Comment (maelvls) [reply pkg/venafi/tpp/connector_test.go]: Yes. I'll make that clearer in the test name.

**Review by inteon — APPROVED** [review]
Thank you for fixing the remaining issues.

File: pkg/venafi/tpp/connector_test.go [hunk 3]
```diff
@@ -528,6 +530,110 @@ func TestRequestCertificateWithValidHours(t *testing.T) {
...
+			givenResetStatus: 400,
+			givenResetBody:   `{"Error": "Reset is not completed. No reset is required for the certificate."}`,
+			expectRequest:    true,
+		},
+		{
+			name:             "should not request a certificate when reset returns an unexpected 401",
+			givenResetStatus: 401,
+			givenResetBody:   `{"error":"session_error","error_description":"Grant rejected scope 'certificate:manage'.}`,
+			expectRequest:    false,
+			expectErr:        "while resetting certificate in case of previous enrollment: 401 Unauthorized, body: ''",
```
Comment (maelvls) [inline pkg/venafi/tpp/connector_test.go:568]: Note: I don't know why the body is `''`. I am expecting the body to be `{"error":"session_error","error_description":"Grant rejected scope 'certificate:manage'.}`, but it seems `request()` doesn't work like that...

Comment (maelvls) [reply pkg/venafi/tpp/connector_test.go]: Ah, that was my fault. I fixed that in aad355f3d712fd870b5b29ead57441d2e949da60.

File: pkg/venafi/tpp/connector.go [hunk 4]
```diff
@@ -678,19 +678,51 @@ func (c *Connector) proccessLocation(req *certificate.Request) error {
...
 	if req.Location != nil {
 		err = c.proccessLocation(req)
 		if err != nil {
 			return
 		}
 	}
+
+	// Let's unconditionally reset any prior pending or failed enrollment. This
+	// doesn't affect the existing certificate if one was already issued. At
+	// first, we wanted to decide whether to reset or not by first retrieving
```
Comment (JigarAtVenafi) [inline pkg/venafi/tpp/connector.go:696]: Hi , I am not sure if I understand the rational behind reseting it every time.  This is edge case we are trying to solve here. 
I would expect Reset to only happen if certificate request is failing with certain error. This seems like an unnecessary API call every time when majority of time it wont be needed. 

Request Cert  --> If Failed with Error --> Reset --> Request again one more time. 
Request cert --> if Good --> Return

File: pkg/venafi/tpp/connector_test.go [hunk 5]
```diff
@@ -28,13 +28,15 @@ import (
...
 	"net/http"
+	"net/http/httptest"
 	"net/url"
 	"os"
 	"path/filepath"
 	"reflect"
 	"regexp"
 	"strconv"
 	"strings"
+	"sync/atomic"
```
Comment (JigarAtVenafi) [inline pkg/venafi/tpp/connector_test.go:39]: This need to be added in go dependencies. 
Jenkisn build is failing

`# github.com/Venafi/vcert/v4/pkg/venafi/tpp [github.com/Venafi/vcert/v4/pkg/venafi/tpp.test]

pkg/venafi/tpp/connector_test.go:579:116: undefined: atomic.Bool

pkg/venafi/tpp/connector_test.go:580:23: undefined: atomic.Bool

FAIL	github.com/Venafi/vcert/v4/pkg/venafi/tpp [build failed]

FAIL`

**Review by JigarAtVenafi — CHANGES_REQUESTED** [review]

**Review by rvelaVenafi — CHANGES_REQUESTED** [review]
There are too many unnecessary calls to reset if set unconditionally. 
The logic to reset based on the request response is better and not more complex than this PR. 
Just move this code to a separate function and validate the request response when statusCode == 500 and msg is the same as the expected JSON blob. 
You can even wrap the current logic into another function with a boolean parameter

```
func (c *Connector) RequestCertificate(req *certificate.Request) (requestID) (requestID string, err error) {
        // first try with reset
        id, err := RequestCertificateWithReset(req, true)
        if err != nil && err == CertResetError{
                 // Second try with no reset
                 id, err := RequestCertificateWithReset(req, false)
        }
}
```

Then have the actual code be moved to a second internal function:

```
func (c *Connector) RequestCertificateWithReset(req *certificate.Request, reset bool) (requestID) (requestID string, err error) {
    // Current RequestCertificate logic here
   
   	requestID, err = parseRequestResult(statusCode, status, body)
	if err != nil {
                 if reset {
		        //invoke reset logic here
                         err = resetCertificateResource(certDN, false)
                         if err != nil {
                                 // error that happened during reset 
                                 return "", err
                         }
                         // This error indicates the caller function that request failed because of 500 and a reset was attempted
                         return "", CertResetError       
                 }
                 // return normal error
                 return "", err 
	}
}
```

This way we can minimize the number of extra calls required for this use case

**Comment by maelvls** [comment]
Thanks for the remarks. The choice of "resetting unconditionally" was motivated by (1) no benefit performance-wise when using the approach "reset only when failing", and (2) implementing "reset only when failing" is very tricky due to the number of undocumented corner cases. I documented the research that led to these results in the PR description.

> There are too many unnecessary calls to reset if set unconditionally.

What do you mean by unnecessary? This HTTP call doesn't seem to be costly (cf. the table with the duration in the PR description). Its cost is comparable to `POST /retrieve`.

File: pkg/venafi/tpp/connector.go (hunk 4 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:696]: I realize I haven't documented the reason why I am not first requesting, and then resetting if that fails.

The reason it is not possible is that `POST  /request` always succeeds as long as the given CSR or certificate parameters are valid. Calling `POST /request` doesn't allow you to know whether `POST /reset` needs to be called or not.

Does that make sense?

File: pkg/venafi/tpp/connector_test.go (hunk 5 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector_test.go:39]: Hmm... `sync/atomic` comes from the Go standard library, which means it doesn't need to be referenced in `go.mod`. 

I will remove the use of `atomic.Bool`.

File: pkg/venafi/tpp/connector.go (hunk 4 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:696]: Since we settled on "Solution 3", I will resolve this conversation.

**Comment by maelvls** [comment]
I finished implementing [Solution 3: request, then reset if retrieve returns “Click Retry” or “WebSDK CertRequest”](https://hackmd.io/@maelvls/design-reset-cert-when-requesting#Solution-3-request-then-reset-if-retrieve-returns-%E2%80%9CClick-Retry%E2%80%9D-or-%E2%80%9CWebSDK-CertRequest%E2%80%9D). Please take another look.

File: pkg/venafi/tpp/connector.go [hunk 6]
```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+			// Continue below.
+		default:
+			return nil, fmt.Errorf("unexpected response while retrieving: %d: %s", statusCode, body)
 		}
+
 		if req.Timeout == 0 {
-			return nil, endpoint.ErrCertificatePending{CertificateID: req.PickupID, Status: retrieveResponse.Status}
+			return nil, endpoint.ErrCertificatePending{CertificateID: req.PickupID, Status: resp.Status}
 		}
 		if time.Now().After(startTime.Add(req.Timeout)) {
```
Comment (inteon) [inline pkg/venafi/tpp/connector.go:1304]: Can you also fix this calculation and add the 2 seconds sleep period?
```suggestion
		if time.Now().Add(2 * time.Second).After(startTime.Add(req.Timeout)) {
```

File: pkg/venafi/tpp/connector.go [hunk 7]
```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %w", err)
+			}
+			if statusCode != http.StatusOK {
+				return nil, fmt.Errorf("while resetting certificate due to a past failed enrollment: %s, body: '%s'", status, body)
+			}
+
+			continue
+		case statusCode == http.StatusInternalServerError && resp.Status != "":
+			return nil, fmt.Errorf("enrollment failed at stage %d: %s", resp.Stage, resp.Status)
+		case statusCode == http.StatusOK:
```
Comment (inteon) [inline pkg/venafi/tpp/connector.go:1283]: Instead of checking for `http.StatusOK` or `http.StatusAccepted`, can we just check `resp.CertificateData != ""` identical to the original code?

File: pkg/venafi/tpp/connector.go [hunk 8]
```diff
@@ -1232,25 +1244,65 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+		switch {
+		// After many tests, we found that the only way to know if the current
+		// certificate request is stuck due to an old failed enrollment is to
+		// look for "WebSDK CertRequest" or "Fix any errors, and then click
+		// Retry." in the retrieve response. We only check this on the first
+		// retrieve because we know that these two messages always disappear
+		// after resetting.
+		case statusCode == http.StatusInternalServerError && retrieveCount == 1 &&
+			(resp.Status == "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry." ||
+				resp.Status == "WebSDK CertRequest Module Requested Certificate"):
```
[the remark this answers is not in the document]
Comment (inteon) [reply pkg/venafi/tpp/connector.go:1268]: (this counter would have to be `==0` for a retry to happen)

File: pkg/venafi/tpp/connector.go (hunk 7 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1283]: I can replace `http.StatusOK` with `resp.CertificateData != ""`, but I would not be able to also do that for `http.StatusAccepted`, since I would fall into the `default` case (which results in an error).

File: pkg/venafi/tpp/connector.go (hunk 6 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1304]: I assumed that fixing this "bug" may break people, that's why I did not change it. But now that I think about it, I can't find any example of breakage that may happen if we fix this.

I'll do the change.

File: pkg/venafi/tpp/connector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1268]: For context, Tim was getting that same message when one of Windows services was down, but that certificate wasn't in error.

@tr1ck3r In https://github.com/Venafi/vcert/issues/239#issuecomment-1242109588, you mentioned that TPP "will" return that body for a certificate object that was in error:

> It will have the following body after making an API request for a certificate object that was in error.
>
> ```json
> {
>   "Stage": 500, 
>   "Status": "WebSDK CertRequest Module Requested Certificate"
> }
> ```

Does this come from your knowledge of the TPP codebase, or is this something that you have observed over time?

File: pkg/venafi/tpp/connector.go (hunk 7 above)
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1286]: Looking at the [`POST /retrieve` documentation](https://docs.venafi.com/Docs/current/TopNav/Content/SDK/WebSDK/r-SDK-POST-Certificates-retrieve.php), it seems like 200 is the only HTTP code for which `CertificateData` will be present. For 202, the only documented JSON fields are `Status` and `Stage`, `CertificateData` isn't meant to be returned on a 202.

But I'd prefer someone confirming that 202 never returns a `CertificateData`.

File: pkg/venafi/tpp/connector.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1264]: **tl;dr:** I'll go with the behavior we discussed, since "WebSDK CertRequest" means that there is a failed past enrollment in "sane" TPP instances.

**Context:** We had some back and forth with Tim. Since he has seen a case where TPP 22.04 would return the message "WebSDK CertRequest Module Requested Certificate" even though no prior enrollment existed, I wanted to investigate. Tim mentioned some Windows service was down that was causing the message. But I was not able to reproduce this issue. I assume it is a case of "misbehaving TPP" and won't happen in production TPP instances.

Tim suggested that we skip the "WebSDK CertRequest" messages until we find a "Click Retry" message.

I did a series of 100 tests with a TPP 20.1 server. Each time, I make sure the certificate is failing, then I request a certificate, and then I retrieve.

Case 1 occurs 60% of the time:

```
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
```

Case 2 occurs 30% of the time:

```
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
{"Stage":500,"Status":"This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry."}
```

Case 3 occurs 10% of the time:

```
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
{"Stage":500,"Status":"WebSDK CertRequest Module Requested Certificate"}
```

Results:
- We cannot only solely on "Click Retry": we would miss 10% of the cases that we are trying to fix (cf. case 3).
- We cannot, in my mind, skip the "WebSDK CertRequest" messages until we see the "Click Retry" message, because it means we would miss 10% of the cases that we were trying to fix. Another minor problem with this approach is that we would end up waiting until the end of the timeout, and not reset.

File: pkg/venafi/tpp/tpp.go [hunk 9]
```diff
@@ -624,25 +630,6 @@ func parseRequestData(b []byte) (data certificateRequestResponse, err error) {
 	return
 }
 
-func parseRetrieveResult(httpStatusCode int, httpStatus string, body []byte) (certificateRetrieveResponse, error) {
```
Comment (marcos-albornoz) [inline pkg/venafi/tpp/tpp.go:627]: Is it possible to keep these functions? that is with the idea to keep the approach to have the parsing of the results in dedicated functions

Comment (maelvls) [reply pkg/venafi/tpp/tpp.go]: I reverted my changes and made sure `parseRetrieveResult` stays. Please take another look. Thanks!

**Review by marcos-albornoz — APPROVED** [review]

**Review by luispresuelVenafi — APPROVED** [review]

File: pkg/venafi/tpp/connector.go [hunk 10]
```diff
@@ -678,19 +678,51 @@ func (c *Connector) proccessLocation(req *certificate.Request) error {
...
+	// When the response to /reset is 400, we just ignore it; here are the three
+	// expected responses:
+	//
+	// | Happy responses                         | Description                |
+	// |-----------------------------------------|----------------------------|
+	// | 200 {"ProcessingResetCompleted": true}  | Success                    |
+	// | 400 {"Error":"CertificateDN error..."}  | Certificate DN not found   |
+	// | 400 {"Error":"No reset is required..."} | No enrollment to be reset  |
+	certDN := getCertificateDN(c.zone, req.Subject.CommonName)
+	statusCode, status, body, err := c.request("POST", urlResourceCertificateReset, struct {
```
[the remark this answers is not in the document]
Comment (maelvls) [reply pkg/venafi/tpp/connector.go]: (from https://github.com/Venafi/vcert/pull/269#issuecomment-1321795062) It is not possible to use `POST /request` as a mean to know whether this certificate should be reset. The reason it is not possible is that `POST  /request` always succeeds as long as the given CSR or certificate parameters are valid. Calling `POST /request` doesn't allow you to know whether `POST /reset` needs to be called or not. I have elaborated a bit on this in https://github.com/Venafi/vcert/pull/269#discussion_r1027814433.

File: pkg/venafi/tpp/connector.go [hunk 11]
```diff
@@ -1256,6 +1279,27 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+func shouldReset(err error) bool {
+	if err == nil {
+		return false
+	}
+
+	if !strings.Contains(err.Error(), "Status: 500") {
+		return false
+	}
+
+	if strings.Contains(err.Error(), "This certificate cannot be processed while it is in an error state. Fix any errors, and then click Retry.") ||
```
Comment (JigarAtVenafi) [inline pkg/venafi/tpp/connector.go:1295]: This is good. It is bit thorough checking. 
Just suggestion may be we can only search strings.Contains("error state") in first condition.  It might not be require if TPP will not be changing error.  No blocker.

Comment (maelvls) [reply pkg/venafi/tpp/connector.go:1295]: I agree, we could use a shorter substring to prevent potential future breakages if TPP changes this error message. I can change that in a later PR (this PR has being going for a while now).

**Review by rvelaVenafi — APPROVED** [review]

**Comment by maelvls** [comment]
Announcement: this PR made it into vcert v4.23.0.

**Comment by maelvls** [comment]
This improvement was reverted in 4.24.0. You can read https://github.com/Venafi/vcert/issues/273#issuecomment-1556938953 to learn more.
